bubuku/features/rebalance_by_size.py

```diff
@@ -0,0 +1,116 @@
+            _LOG.info("Starting rebalance by size")
+            return RebalanceBySizeChange(self.zk)
+        return None
+
+    def __is_data_imbalanced(self) -> bool:
+        return False  # todo: not implemented yet
+
+
+class GenerateDataSizeStatistics(Check):
+    def __init__(self, zk: Exhibitor, broker: BrokerManager, cmd_helper: CmdHelper, kafka_log_dirs):
+        super().__init__(check_interval_s=1800)
+        self.zk = zk
+        self.broker = broker
+        self.cmd_helper = cmd_helper
+        self.kafka_log_dirs = kafka_log_dirs
+
+    def check(self):
+        if self.broker.is_running_and_registered():
+            _LOG.info("Generating data size statistics")
+            try:
+                self.__generate_stats()
+                _LOG.info("Data size statistics successfully written to zk")
+            except Exception:
+                _LOG.warn("Error occurred when collecting size statistics", exc_info=True)
+        return None
+
+    def __generate_stats(self):
+        topic_stats = self.__get_topics_stats()
+        disk_stats = self.__get_disk_stats()
+        stats = {"disk": disk_stats, "topics": topic_stats}
+        self.__write_stats_to_zk(stats)
+
+    def __get_topics_stats(self):
+        topic_stats = {}
+        for log_dir in self.kafka_log_dirs.split(","):
+            topic_dirs = self.cmd_helper.cmd_run("du -k -d 1 {}".format(log_dir)).split("\n")
+            for topic_dir in topic_dirs:
+                self.__parse_dir_stats(topic_dir, log_dir, topic_stats)
+        return topic_stats
+
+    @staticmethod
+    def __parse_dir_stats(dir, log_dir, topic_stats):
+        """
+        Parses the "du" tool single line output and writes result to topic_stats json
+        :param dir: the string to be parsed; example: "45983\t/tmp/kafka-logs/my-kafka-topic-0"
+        :param log_dir: the kafka log directory name itself
+        :param topic_stats: the json which will be populated
+        """
+        dir_data = dir.split("\t")
+        if len(dir_data) == 2 and dir_data[1] != log_dir:
+            size_kb, dir_name = tuple(dir_data)
+            tp_name = dir_name.split("/")[-1]
+            tp_parts = tp_name.rsplit("-", 1)
+            if len(tp_parts) == 2:
+                topic, partition = tuple(tp_parts)
+                if topic not in topic_stats:
+                    topic_stats[topic] = {}
+                topic_stats[topic][partition] = int(size_kb)
+
+    def __get_disk_stats(self):
+        disks = self.cmd_helper.cmd_run("df -k | tail -n +2 |  awk '{ print $3, $4 }'").split("\n")
+        total_used = total_free = 0
+        for disk in disks:
+            parts = disk.split(" ")
+            if len(parts) == 2:
+                used, free = tuple(parts)
+                total_used += int(used)
+                total_free += int(free)
+        return {"used": total_used, "free": total_free}
+
+    def __write_stats_to_zk(self, stats):
```

Comment:
Well... It may happend that the maximum size is reached, from: https://zookeeper.apache.org/doc/r3.1.2/api/org/apache/zookeeper/ZooKeeper.html
```
The maximum allowable size of the data array is 1 MB (1,048,576 bytes). Arrays larger than this will cause a KeeperExecption to be thrown.
```